File: Design-Patterns-Using-Cpp/State/StatePattern.cpp
#include <iostream>
#include <string>

class BulbControl;

class BulbState {
public:
    virtual void flip(BulbControl& bc) = 0;
    virtual std::string toString() const = 0;
    virtual ~BulbState() {}
};

class On : public BulbState {
public:
    void flip(BulbControl& bc) override;
    std::string toString() const override;
};

class Off : public BulbState {
public:
    void flip(BulbControl& bc) override;
    std::string toString() const override;
};

class BulbControl {
private:
    BulbState* current;

public:
    BulbControl() {
        this->current = new Off();
    }

    void setState(BulbState* state) {
        delete this->current;
        this->current = state;
    }

    void flip() {
        this->current->flip(*this);
    }

    std::string toStringState() const {
        return this->current->toString();
    }

    ~BulbControl() {
        delete this->current;
    }
};



void On::flip(BulbControl& bc) {
    bc.setState(new Off());
}

std::string On::toString() const {
    return "On";
}

void Off::flip(BulbControl& bc) {
    bc.setState(new On());
}

std::string Off::toString() const {
    return "Off";
}

// Client code
int main() {
    BulbControl c;
    c.flip();
    std::cout << c.toStringState() << std::endl;
    c.flip();
    std::cout << c.toStringState() << std::endl;

    return 0;
}

/*
On
Off
*/

File: Design-Patterns-Using-Kotlin/State/StatePattern.kt
interface BulbState {
    fun flip(bc: BulbControl)
    override fun toString(): String
}

class BulbControl {
    private var current: BulbState

    init {
        this.current = Off()
    }

    fun setState(state: BulbState) {
        this.current = state
    }

    fun flip() {
        current.flip(this)
    }

    fun toStringState(): String {
        return current.toString()
    }
}

class On : BulbState {
    override fun flip(bc: BulbControl) {
        bc.setState(Off())
    }

    override fun toString(): String {
        return "On"
    }
}

class Off : BulbState {
    override fun flip(bc: BulbControl) {
        bc.setState(On())
    }

    override fun toString(): String {
        return "Off"
    }
}

// Client code.
fun main() {
    val c = BulbControl()
    c.flip()
    println(c.toStringState())
    c.flip()
    println(c.toStringState())
}

/*
On
Off
*/